proposals/0000-allow-constraint-type-synonyms-in-deriving-heads.md

```diff
@@ -0,0 +1,204 @@
+   split such declarations into multiple.
+
+This should work the same for any kind of instances:
+stanalone derived instances with or without explicit context,
+non-standalone derived instances
+and user-defined instances.
+
+Thus proposal do not introduce any changes to instance semantics,
+only allowed syntax and user error messages would be improved.
+
+## Examples
+
+### Common derived instances usecase
+
+```haskell
+type ToFromJSON x = (ToJSON x, FromJSON x)
+
+-- Allowed by this proposal and working same as example above
+data MyData = MkMyData deriving (ToFromJSON)
+
+-- Allowed by this proposal
+derive instance ... => (ToFromJSON MyData)
+
+-- Should work the same with syntax allowed before proposal
+derive instance ... => ToJSON MyData
+derive instance ... => FromJSON MyData
+```
+
+Non-standalone instances do not have explicit params,
+so rewriting implementation may be little different.
+But they have just the same samantics as standalone instances with holes,
+covered by examples before, so could be handled the same.
+
+### Tuple of classes instance usecase
+
+Example from @VitWW :
+
+```haskell
+type Num x = (Arithm x, Abs x)
+type Arithm x = (Add x, Sub x, Mult x)
+
+data MyData
+
+instance Num MyData where
+    (+) = ...
+    (*) = ...
+
+-- same as
+instance Add MyData where
+    (+) = ...
+
+instance Mult MyData where
+    (*) = ...
+```
+
+## Effect and Interactions
+
+There is already extension named `TypeSynonymInstances`,
+but it does not cover constraint synonym case.
+Probably, constraint synonyms should be guarded by this extension as well
+and documentation updated on that.
+
+It may be, that some linters depend on current syntaxic restrictions,
+but I am not aware on any specific cases.
+
+## Costs and Drawbacks
+
+Main concern is that rewriting implementations may make errors less clear.
+I propose to mitigate if this happens by adding context to errors,
+same as it done for instances resolving now.
+
+## Backward Compatibility
+
+No breakage. Part of this even was working before.
+
+## Alternatives
+
+For deriving instances there are two alternatives:
+
+1. Use CPP preprocessor variable, which breaks type-safety,
+   may lead to surprising error messages and is less composable
+2. Use Template Haskell, which have its own problems and is less simple for beginners
+
+For user-defined instances one may define type class instead of synonym:
+
+```haskell
+class (ToJSON x, FromJSON x) => ToFromJSON x
+
+instance (ToJSON x, FromJSON x) => ToFromJSON x
+```
+
+While this trick works, Haskell begginers may be not aware of it.
+Also it makes less clear API, because `ToFromJSON` is should be same as
+typeclass tuple only by syntaxic convention.
+
+## Unresolved Questions
+
+The only question is, which implementation approach to use.
+
+I think this question is not blocking,
+because all implementation paths lead to same semantics,
+only potential difference may be slightly less-clear type errors.
```

Comment:
```suggestion
because all implementation paths lead to same semantics.
A potential difference in implementation paths could lead to slightly sub-optimal errors messages.
```